### DISCUSSION TITLE
SHACL Problem : Shape with multiple types: <http://www.w3.org/ns/shacl#PropertyShape>, <http://www.w3.org/ns/shacl#NodeShape>

### DISCUSSION THREAD

**Discussion by volkerjaenisch** [eclipse-rdf4j/rdf4j#4287] [General]
Dear RDF4J Users!

Currently we am implementing SHACL for an OpenData portal (http://datenadler.de).
We are Python programmers so we planed to go with pySHACL. But our Plan B is RDF4J which we are currently using as our backend.

Testing our Plan B we stumbled over the following exception of RDF4J reading a SHAPE file.

```
javax.servlet.ServletException: org.eclipse.rdf4j.repository.RepositoryException: Shape with multiple types: <http://www.w3.org/ns/shacl#PropertyShape>, <http://www.w3.org/ns/shacl#NodeShape>
	org.eclipse.rdf4j.workbench.proxy.WorkbenchServlet.handleRequest(WorkbenchServlet.java:160)
	org.eclipse.rdf4j.workbench.proxy.WorkbenchServlet.service(WorkbenchServlet.java:112)
	org.eclipse.rdf4j.workbench.proxy.WorkbenchGateway.service(WorkbenchGateway.java:117)
	org.eclipse.rdf4j.workbench.base.AbstractServlet.service(AbstractServlet.java:129)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	org.eclipse.rdf4j.workbench.proxy.CacheFilter.doFilter(CacheFilter.java:64)
	org.eclipse.rdf4j.workbench.proxy.CookieCacheControlFilter.doFilter(CookieCacheControlFilter.java:56)

Root Cause

org.eclipse.rdf4j.repository.RepositoryException: Shape with multiple types: <http://www.w3.org/ns/shacl#PropertyShape>, <http://www.w3.org/ns/shacl#NodeShape>
	org.eclipse.rdf4j.http.client.SPARQLProtocolSession.execute(SPARQLProtocolSession.java:1095)
	org.eclipse.rdf4j.http.client.SPARQLProtocolSession.executeNoContent(SPARQLProtocolSession.java:1049)
	org.eclipse.rdf4j.http.client.RDF4JProtocolSession.upload(RDF4JProtocolSession.java:1103)
	org.eclipse.rdf4j.http.client.RDF4JProtocolSession.upload(RDF4JProtocolSession.java:928)
	org.eclipse.rdf4j.http.client.RDF4JProtocolSession.upload(RDF4JProtocolSession.java:919)
	org.eclipse.rdf4j.repository.http.HTTPRepositoryConnection.add(HTTPRepositoryConnection.java:447)
	org.eclipse.rdf4j.workbench.commands.AddServlet.add(AddServlet.java:94)
	org.eclipse.rdf4j.workbench.commands.AddServlet.doPost(AddServlet.java:53)
	org.eclipse.rdf4j.workbench.base.TransformationServlet.service(TransformationServlet.java:98)
	org.eclipse.rdf4j.workbench.base.AbstractServlet.service(AbstractServlet.java:129)
	org.eclipse.rdf4j.workbench.proxy.ProxyRepositoryServlet.service(ProxyRepositoryServlet.java:100)
	org.eclipse.rdf4j.workbench.proxy.WorkbenchServlet.service(WorkbenchServlet.java:215)
	org.eclipse.rdf4j.workbench.proxy.WorkbenchServlet.handleRequest(WorkbenchServlet.java:137)
	org.eclipse.rdf4j.workbench.proxy.WorkbenchServlet.service(WorkbenchServlet.java:112)
	org.eclipse.rdf4j.workbench.proxy.WorkbenchGateway.service(WorkbenchGateway.java:117)
	org.eclipse.rdf4j.workbench.base.AbstractServlet.service(AbstractServlet.java:129)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	org.eclipse.rdf4j.workbench.proxy.CacheFilter.doFilter(CacheFilter.java:64)
	org.eclipse.rdf4j.workbench.proxy.CookieCacheControlFilter.doFilter(CookieCacheControlFilter.java:56)

```

Can someone please explain what is going wrong.

The shape file comes from the EU  
https://github.com/SEMICeu/DCAT-AP/blob/master/releases/2.1.1/dcat-ap_2.1.1_shacl_shapes.ttl

pySHACL reads and uses this file with no complains. I also checked graphDB (to make sure my RDF4J is installed correctly) with the exact same outcome.

Any help appreciated

Cheers,
Volker

**Comment by barthanssens** [comment]
Hi Volker,

could you tell us which RDF4J version you're using ? 
Thanks

Bart

**Comment by volkerjaenisch** [comment]
Dear Bart!

Am 19.11.22 um 22:15 schrieb Bart Hanssens:
>
> Hi Volker,
>
> could you tell us which RDF4J version you're using ?
> Thanks
>
I am using the code from yesterday :-)

Application Information
Application Name    RDF4J Workbench
Version    4.2.1

Cheers,
Volker


-- 
=========================================================
    inqbus Scientific Computing    Dr.  Volker Jaenisch
    Hungerbichlweg 3               +49 (8860) 9222 7 92
    86977 Burggenhttps://inqbus.de
=========================================================

**Comment by volkerjaenisch** [comment]
Dear Bart!


    Application Information

Application name: 	RDF4J Server
Version: 	4.2.1
Data directory: 	/var/lib/tomcat/.RDF4J/server


    Runtime Information

Operating System: 	Linux 5.10.0-19-amd64 (amd64)
Java Runtime: 	Debian OpenJDK 64-Bit Server VM 11.0.16
Process User: 	tomcat


    Memory

Percentage in use: 	4 %
Used: 	320 MB
Maximum: 	7.994 MB

-- 

=========================================================
    inqbus Scientific Computing    Dr.  Volker Jaenisch
    Hungerbichlweg 3               +49 (8860) 9222 7 92
    86977 Burggenhttps://inqbus.de
=========================================================

**Comment by barthanssens** [comment]
Thanks, I'll look into it in the afternoon

**Reply by volkerjaenisch** [reply]
@barthanssens 
Thank you so much!

**Comment by barthanssens** [comment]
Looks like RDF4J has an issue with `sh:name`:  if I copy the DCAT-AP file and remove the `sh:name` property from all the shapes, the SHACL file is loaded without error

**Comment by barthanssens** [comment]
Arguably this is not a bug, since, according to the SHACL ontology (https://github.com/w3c/data-shapes/blob/gh-pages/shacl/shacl.ttl), `sh:name` is (only) a property of a `sh:PropertyShape`

While the DCAT-AP shacl file uses it on `sh:NodeShape`, effectively turning it into a sh:PropertyShape as well (hence the exception being thrown)

**Reply by volkerjaenisch** [reply]
Ah! 
Thank you for the explanation. I understand the cause of the exception, now.
But how to proceed from here?

I may open a bugreport at DCAT-AP and address the ontology domain problem:

```
sh:name
	a rdf:Property ;
	rdfs:label "name"@en ;
	rdfs:comment "Human-readable labels for the property in the context of the surrounding shape."@en ;
	rdfs:domain sh:PropertyShape ;
	# range: xsd:string or rdf:langString
	rdfs:isDefinedBy sh: .
```
But I guess that the answer will be: "A SHACL entity given a sh:name then has an additional type,  so what? Problem of your engine" :-). 

And this argumentation is not completely nonsense. Is it justified to throw an exception in this case? Does having two types breaks SHACL. Since two other SHACL processors correctly using this shape file, I would think that this is not the case.

Also the DCAT-AP SHACL shape runs without flaw through the SHACL shape validator:  [https://www.itb.ec.europa.eu/shacl/shacl/upload](https://www.itb.ec.europa.eu/shacl/shacl/upload).

Maybe this exception is really not necessary.

Cheers,
Volker

**Reply by barthanssens** [reply]
Hi @volkerjaenisch 

I've opened an issue https://github.com/SEMICeu/DCAT-AP/issues/245 (so at least people are aware that it might be an issue in at least 1 engine)

While technically RDF4J is correct (IMHO), I do understand that it makes sense that this property is also present on NodeShapes
(this should be fixed by the W3C SHACL working group)

So we can look into a fix/work-around in RDF4J, just trying to find out what caused the issue.

Best regards

Bart

**Reply by volkerjaenisch** [reply]
:thumbsup:

**Comment by barthanssens** [comment]
A workaround could be e.g. something along the lines of :

```
Model rules = Rio.parse(new FileReader("dcat-ap-221.ttl"), RDFFormat.TURTLE, (Resource) null);
rules.remove(null, SHACL.NAME, null);

connection.begin();
connection.add(rules, RDF4J.SHACL_SHAPE_GRAPH);
connection.commit();
```

**Reply by volkerjaenisch** [reply]
Yes, that will surely workaround the exception, but on the other Hand the sh:name attributes are quite important for the human readability of the result report.

**Reply by barthanssens** [reply]
Agree :-) But as mentioned, actually the SHACL specification itself should be updated to also allow sh:name on sh:NodeShape (instead of only on sh:PropertyShape) ... 

The DCAT-AP shacl file somewhat incorrectly uses it on a sh:NodeShape
